R/dsl-generate.R

```diff
@@ -27,10 +27,26 @@ dsl_generate <- function(dat) {
 
 dsl_generate_density <- function(dat, env, meta) {
   exprs <- lapply(dat$exprs, dsl_generate_density_expr, meta)
-  body <- c(call("<-", meta[["pars"]], quote(packer$unpack(x))),
+  body_exprs <- c(call("<-", meta[["pars"]], quote(packer$unpack(x))),
             call("<-", meta[["density"]], quote(numeric())),
             exprs,
             call("sum", meta[["density"]]))
+  if (is.null(dat$domain)) {
+    body <- rlang::call2("{", !!!body_exprs)
+  } else {
+    if (nrow(dat$domain) == 1) {
+      domain_min <- dat$domain[[1]]
+      domain_max <- dat$domain[[2]]
+      in_domain <- bquote(x >= .(domain_min) && x <= .(domain_max))
+    } else {
+      domain_min <- rlang::call2("c", !!!unname(dat$domain[, 1]))
+      domain_max <- rlang::call2("c", !!!unname(dat$domain[, 2]))
+      in_domain <- bquote(all(x >= .(domain_min) & x <= .(domain_max)))
+    }
+    body <- call(
+      "{",
+      call("if", in_domain, rlang::call2("{", !!!body_exprs), call("{", -Inf)))
```

Comment:
Note this is the opposite logic to the function case, might be worth inverting it